packages/houdini-utils/src/util/featureDetect.ts

```diff
@@ -0,0 +1,94 @@
+import { addModule } from './addModule';
+
+declare global {
+  interface Document {
+    mozSetImageElement?: (
+      imageElementId: string,
+      imageElement: HTMLCanvasElement
+    ) => void;
+    getCSSCanvasContext?: (
+      contextId: string,
+      name: string,
+      width: number,
+      height: number
+    ) => CanvasRenderingContext2D;
+  }
+
+  interface CSS {
+    paintWorklet: {
+      addModule: (url: string) => Promise<void>;
+    };
+  }
+}
+
+export type FeatureDetectFn = () => boolean;
+export type AsyncFeatureDetectFn = () => Promise<boolean>;
+
+/**
+ * Test if the APIs neccessary for the Firefox fallback exist.
+ *
+ * @returns `true` if the browser supports the necessary APIs, `false` otherwise.
+ */
+export const hasMozElement: FeatureDetectFn = () => {
+  return typeof document?.mozSetImageElement === 'function';
+};
+
+/**
+ * Test if the APIs necessary for the Safari fallback exist.
+ * @returns `true` if the browser supports the necessary APIs, `false` otherwise.
+ */
+export const hasWebkitCanvas: FeatureDetectFn = () => {
+  return typeof document?.getCSSCanvasContext === 'function';
+};
+
+/**
+ * Test if CSS Houdini APIs are availabe.
+ * @see canUseHoudini() for a more robust test.
+ * @returns `true` if the browser supports necessary Houdini APIs, `false` otherwise.
+ */
+export const hasHoudini: FeatureDetectFn = () => {
+  return (
+    typeof window !== 'undefined' &&
+    window.CSS &&
+    'paintWorklet' in CSS &&
+    'registerProperty' in CSS
+  );
+};
+
+let canUseHoudiniCache: boolean | undefined = undefined;
+/**
+ * A more robust test for CSS Houdini APIs. This function also verifies that a simple
+ * PaintWorklet can be loaded.
+ *
+ * Note that this check will add a PaintWorklet as a blob URL, apps with CSP policies will always fail this check.
+ * @see hasHoudini
+ * @returns `Promise<true>` if the browser supports necessary Houdini APIs, `Promise<false>` otherwise.
+ */
+export const canUseHoudini: AsyncFeatureDetectFn = async () => {
```

Comment:
I'd like to eliminate this function. It's async which introduces a race condition when loading the PaintWorklet and will always fail for apps with modest CSP settings -- preventing them from using the Houdini feature they might otherwise be able to use.

Instead I think we should rely on the synchronous `hasHoundini()` as a baseline test of whether the necessary APIs exist and rely on the result of `addModule()` to potentially use the fallback in the case a Houdini-supporting browser fails to load the `PaintWorklet` for some reason.